[ie/neteasemusic] Use new API for url extraction

hafeoz opened the pull request:
### Description of your *pull request* and other information

<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

Netease Music now seems to be using a new API (discussed in https://github.com/yt-dlp/yt-dlp/issues/9196#issuecomment-2185285724 and https://github.com/yt-dlp/yt-dlp/pull/9197#discussion_r1650157304), and premium music qualities can only be obtained this way. This PR updates the API based on [other](https://github.com/HyPlayer/NeteaseCloudMusicAPI/blob/63c7f30975bdf88d19252f3cd1bbad8ea97eb450/NeteaseCloudMusicApi/CloudMusicApiProviders.cs#L789) [similar](https://github.com/Suxiaoqinx/Netease_url/blob/29b47c53372a744184b1d896e656fef5e4a8f9b1/main.py#L114) projects.

Fixes #9196 #10239


<details open><summary>Template</summary> <!-- OPEN is intentional -->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))
</details>

hafeoz commented:
@timi137137 Can you test if the API used in this PR is working for VIP accounts? I'm currently oversea and doesn't have Netease VIP. 你能试试这个PR获取到的音频正常吗？我没有网易云VIP。如果可以的话能发一下
```shell
git clone https://github.com/hafeoz/yt-dlp.git && cd yt-dlp
python yt_dlp/__main__.py --list-formats --cookies cookies.txt -vU 'https://music.163.com/song?id=2058263030'
```
的输出确认一下所有的音质都被获取到了吗？

timi137137 commented:
@hafeoz 完美解决，太开心了。七种音质全都抓取到了并且大小也没错
Perfect solution！
We got all seven qualities and the right size.

![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/3687aa3d-029c-4361-b232-4ce6807bb81e)

bashonly commented on yt_dlp/extractor/neteasemusic.py:
```diff
@@ -66,27 +66,22 @@ def _download_eapi_json(self, path, video_id, query_body, headers={}, **kwargs):
                         'url': song_url,
-                        'format_id': song_format,
-                        'asr': traverse_obj(details, ('sr', {int_or_none})),
+                        'format_id': song_level,
```
should also add this so yt-dlp knows it's an audio-only format
```suggestion
                        'format_id': song_level,
                        'vcodec': 'none',
```

hafeoz replied:
Got it thanks!

dirkf commented:
Test `test_NetEaseMusic` is failing from UK -> no media links, geo-restriction?
Test `test_NetEaseMusic_1` is missing `'description': '[00:05.00]纯音乐，请欣赏\n'`.
Test `test_NetEaseMusic_2` fails on the file MD5.
The skipped test `test_NetEaseMusic_4` and also `test_NetEaseMusic` get JSON like this in `_call_player_api()` (I didn't check this in skipped test `test_NetEaseMusic_3`):
```json 
{
    "code": 200,
    "data": [{
        ...,
        "code": 404,
        ...,
    }]
}
```
Maybe this is an extra-stupid API that returns 200 in HTTP and top-level JSON even when it means 404 and the item is actually expired? Or this 404 could just mean "not saying".

timi137137 commented:
@dirkf I guess you'll need [this](https://github.com/UnblockNeteaseMusic)
After all, music is copyrighted, and regional restrictions certainly exist

dirkf commented:
I'm neutral about any restrictions, but not about detecting when they apply and having that reflected in tests (thx bashonly)!

Presumably `test_NetEaseMusic_2` was finding and selecting a FLAC format.

timi137137 commented:
The test passed.
Here are the test details

<details> <summary>Test Details</summary>
test_NetEaseMusic_1： 
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/d5f6140c-eee1-48b4-9b60-ac75c6238b8e)
The first one was taken offline

---
test_NetEaseMusic_2:
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/05ca9376-c798-4d31-9a67-728b63b55482)
There are actually only two kinds of sound quality

---
test_NetEaseMusic_3:
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/71d4cd4e-50ce-4926-bf29-3f91a06bface)
Correct output

---
test_NetEaseMusic_4:
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/e9a42fd6-1055-45f3-b806-9e149c53cd1d)
Correct output

---
test_NetEaseMusic_5:
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/356301ee-3256-4aba-80b3-563c4cf9d96f)
Correct output
</details>

timi137137 commented:
Thank you very much [hafeoz](https://github.com/hafeoz), [bashonly](https://github.com/bashonly), and [dirkf](https://github.com/dirkf) 
Thank you so much for helping me improve my experience.


Some things to note:
1. Not all songs are available in flac format, because many songs are too old for companies to update one by one.
2. Not all songs can get immersive surround sound, the conditions for this are still being found, but it is certain that it will not affect users, because jymaster is the next best sound quality(For now)

bashonly commented:
There is no difference in the player API response for a song that has been removed vs. one that is not available due to geo-restriction. 

Because of this, and because of what dirkf pointed out about the top-level status code always being 200, the error handling in `_extract_formats` is largely pointless now. Should we just do this?

```diff
diff --git a/yt_dlp/extractor/neteasemusic.py b/yt_dlp/extractor/neteasemusic.py
index db8ce65c7..b46502d73 100644
--- a/yt_dlp/extractor/neteasemusic.py
+++ b/yt_dlp/extractor/neteasemusic.py
@@ -72,7 +72,6 @@ def _call_player_api(self, song_id, level):
             note=f'Downloading song URL info: level {level}')
 
     def extract_formats(self, info):
-        err = 0
         formats = []
         song_id = info['id']
         for song_level in self._LEVELS:
@@ -89,15 +88,9 @@ def extract_formats(self, info):
                             'filesize': ('size', {int_or_none}),
                         }),
                     })
-                elif err == 0:
-                    err = traverse_obj(song, ('code', {int})) or 0
-
         if not formats:
-            if err != 0 and (err < 200 or err >= 400):
-                raise ExtractorError(f'No media links found (site code {err})', expected=True)
-            else:
-                self.raise_geo_restricted(
-                    'No media links found: probably due to geo restriction.', countries=['CN'])
+            self.raise_geo_restricted(
+                'No media links found: probably due to geo restriction.', countries=['CN'])
         return formats
 
     def query_api(self, endpoint, video_id, note):
```

bashonly commented:
Furthermore, we are now making 14 requests for formats extraction. Do we really need the `_is_valid_url()` call? Are there any examples where this is actually helpful? We could cut the number of requests in half if we do this:

```diff
diff --git a/yt_dlp/extractor/neteasemusic.py b/yt_dlp/extractor/neteasemusic.py
index db8ce65c7..71027853a 100644
--- a/yt_dlp/extractor/neteasemusic.py
+++ b/yt_dlp/extractor/neteasemusic.py
@@ -72,32 +72,23 @@ def _call_player_api(self, song_id, level):
             note=f'Downloading song URL info: level {level}')
 
     def extract_formats(self, info):
-        err = 0
         formats = []
         song_id = info['id']
         for song_level in self._LEVELS:
             for song in traverse_obj(self._call_player_api(song_id, song_level), ('data', lambda _, v: url_or_none(v['url']))):
-                song_url = song['url']
-                if self._is_valid_url(song_url, info['id'], 'song'):
-                    formats.append({
-                        'url': song_url,
-                        'format_id': song_level,
-                        'vcodec': 'none',
-                        **traverse_obj(song, {
-                            'ext': ('type', {str}),
-                            'abr': ('br', {self.kilo_or_none}),
-                            'filesize': ('size', {int_or_none}),
-                        }),
-                    })
-                elif err == 0:
-                    err = traverse_obj(song, ('code', {int})) or 0
-
+                formats.append({
+                    'url': song['url'],
+                    'format_id': song_level,
+                    'vcodec': 'none',
+                    **traverse_obj(song, {
+                        'ext': ('type', {str}),
+                        'abr': ('br', {self.kilo_or_none}),
+                        'filesize': ('size', {int_or_none}),
+                    }),
+                })
         if not formats:
-            if err != 0 and (err < 200 or err >= 400):
-                raise ExtractorError(f'No media links found (site code {err})', expected=True)
-            else:
-                self.raise_geo_restricted(
-                    'No media links found: probably due to geo restriction.', countries=['CN'])
+            self.raise_geo_restricted(
+                'No media links found: probably due to geo restriction.', countries=['CN'])
         return formats
 
     def query_api(self, endpoint, video_id, note):
```

timi137137 commented:
> There is no difference in the player API response for a song that has been removed vs. one that is not available due to geo-restriction.
> Because of this, and because of what dirkf pointed out about the top-level status code always being 200, the error handling in _extract_formats is largely pointless now. Should we just do this?

Chinese developers are used to identifying the code that responds to the content, rather than the request status code
Such as:
get something from xxx.xxx/v1/get
in China, there respone http status code will be `200 OK` 
but, respone content be like:
```json
{
  "code": 404,
  "msg": "Not Found",
  "data": {},
}
```
So you should retrieve the status code from within it.
Sorry, I don't know how to change it, so I can only take screenshots
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/57f0196b-852b-435a-8eb1-b392ce9a2bcf)

bashonly commented:
I get the same top-level 200 code and data element 404 codes for both removed songs and geo-blocked songs

timi137137 commented:
This one is normal status return.
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/f0771da8-62d0-4f81-9d22-bc3f3e17239f)

> Furthermore, we are now making 14 requests for formats extraction. Do we really need the _is_valid_url() call? Are there any examples where this is actually helpful? We could cut the number of requests in half if we do this:

I've looked at the implementation of _is_valid_url(), and personally I don't think we need to call this. Because it doesn't return 200 status code if the interface fails. As long as the interface is alive, it returns valid data

bashonly commented:
@hafeoz @timi137137 please test my changes

hafeoz commented:
> @hafeoz @timi137137 please test my changes

Yes I can confirm `NetEaseMusic_all` tests passes for me in UK and manual `--list-formats` works as expected.

I did notice that many other `NetEaseMusic` extractor tests are failing. All `NetEaseMusicProgramIE` tests, for example, is failing (probably due to geo-blocking? Not sure).

<details open><summary>Test results</summary>

```
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusic_all']
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /home/hafeoz/.local/share/hatch/env/virtual/yt-dlp/hPV3QzYK/hatch-test.py3.12/bin/python3
cachedir: .pytest_cache
rootdir: /tmp/yt-dlp
configfile: pyproject.toml
plugins: rerunfailures-14.0, xdist-3.6.1
collecting ... collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusic_all PASSED        [100%]

============================== 1 passed in 7.98s ===============================
```

```

Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicAlbum_all']
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /home/hafeoz/.local/share/hatch/env/virtual/yt-dlp/hPV3QzYK/hatch-test.py3.12/bin/python3
cachedir: .pytest_cache
rootdir: /tmp/yt-dlp
configfile: pyproject.toml
plugins: rerunfailures-14.0, xdist-3.6.1
collecting ... collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicAlbum_all FAILED   [100%]

=================================== FAILURES ===================================
___________________ TestDownload.test_NetEaseMusicAlbum_all ____________________
test/test_download.py:294: in test_template
    getattr(self, test_name)()
test/test_download.py:200: in test_template
    assertGreaterEqual(
test/helper.py:300: in assertGreaterEqual
    self.assertTrue(got >= expected, msg)
E   AssertionError: False is not true : Expected at least 13 in playlist https://music.163.com/#/album?id=133153666, but got only 12
----------------------------- Captured stdout call -----------------------------
[debug] Loaded 1823 extractors
[netease:album] Extracting URL: https://music.163.com/#/album?id=133153666
[netease:album] 133153666: Downloading webpage
[download] Downloading playlist: 桃几的翻唱
[netease:album] Playlist 桃几的翻唱: Downloading 12 items of 12
[debug] The information of all playlist entries will be held in memory
[download] Downloading item 1 of 12
[download] Downloading item 2 of 12
[download] Downloading item 3 of 12
[download] Downloading item 4 of 12
[download] Downloading item 5 of 12
[download] Downloading item 6 of 12
[download] Downloading item 7 of 12
[download] Downloading item 8 of 12
[download] Downloading item 9 of 12
[download] Downloading item 10 of 12
[download] Downloading item 11 of 12
[download] Downloading item 12 of 12
[download] Finished downloading playlist: 桃几的翻唱
=========================== short test summary info ============================
FAILED test/test_download.py::TestDownload::test_NetEaseMusicAlbum_all - Asse...
============================== 1 failed in 2.84s ===============================
```

```
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicSinger_all']
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /home/hafeoz/.local/share/hatch/env/virtual/yt-dlp/hPV3QzYK/hatch-test.py3.12/bin/python3
cachedir: .pytest_cache
rootdir: /tmp/yt-dlp
configfile: pyproject.toml
plugins: rerunfailures-14.0, xdist-3.6.1
collecting ... collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicSinger_all PASSED  [100%]

============================== 1 passed in 4.83s ===============================
```

```
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicList_all']
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /home/hafeoz/.local/share/hatch/env/virtual/yt-dlp/hPV3QzYK/hatch-test.py3.12/bin/python3
cachedir: .pytest_cache
rootdir: /tmp/yt-dlp
configfile: pyproject.toml
plugins: rerunfailures-14.0, xdist-3.6.1
collecting ... collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicList_all PASSED    [100%]

============================== 1 passed in 6.05s ===============================
```

```

Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicMv_all']
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /home/hafeoz/.local/share/hatch/env/virtual/yt-dlp/hPV3QzYK/hatch-test.py3.12/bin/python3
cachedir: .pytest_cache
rootdir: /tmp/yt-dlp
configfile: pyproject.toml
plugins: rerunfailures-14.0, xdist-3.6.1
collecting ... collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicMv_all FAILED      [100%]

=================================== FAILURES ===================================
_____________________ TestDownload.test_NetEaseMusicMv_all _____________________
test/test_download.py:294: in test_template
    getattr(self, test_name)()
test/test_download.py:229: in test_template
    expect_info_dict(self, tc_res_dict, tc.get('info_dict', {}))
test/helper.py:275: in expect_info_dict
    self.assertFalse(
E   AssertionError: {'creators'} is not false : Missing keys in test definition: creators
----------------------------- Captured stdout call -----------------------------
[debug] Loaded 1823 extractors
[netease:mv] Extracting URL: https://music.163.com/#/mv?id=10958064
[netease:mv] 10958064: Downloading mv info
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] 10958064: Downloading 1 format(s): 1080p
[info] Writing video metadata as JSON to: test_NetEaseMusicMv_10958064.info.json
[debug] Invoking http downloader on "http://vodkgeyttp8.vod.126.net/cloudmusic/obj/core/4026923983/1765d168a505e8a29349ada6d039413c.mp4?wsSecret=7520e5afb709dcc5b3f5c6589674eca2&wsTime=1719372330"
[download] Destination: test_NetEaseMusicMv_10958064.mp4
[download] 100% of   10.00KiB in 00:00:00 at 19.51KiB/s  
----------------------------- Captured stderr call -----------------------------

        'info_dict': {
            'id': '10958064',
            'title': '交换余生',
            'description': 'md5:e845872cff28820642a2b02eda428fea',
            'upload_date': '20200916',
            'thumbnail': 'http://p1.music.126.net/dDfogUVQ8A_p3rTEcucNiA==/109951165321342062.jpg',
            'duration': 364,
            'view_count': int,
            'like_count': int,
            'comment_count': int,
        
            'creators': ['林俊杰'],
        },
        
=========================== short test summary info ============================
FAILED test/test_download.py::TestDownload::test_NetEaseMusicMv_all - Asserti...
============================== 1 failed in 2.40s ===============================
```

```

Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicProgram_all']
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /home/hafeoz/.local/share/hatch/env/virtual/yt-dlp/hPV3QzYK/hatch-test.py3.12/bin/python3
cachedir: .pytest_cache
rootdir: /tmp/yt-dlp
configfile: pyproject.toml
plugins: rerunfailures-14.0, xdist-3.6.1
collecting ... collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicProgram_all FAILED [100%]

=================================== FAILURES ===================================
__________________ TestDownload.test_NetEaseMusicProgram_all ___________________
yt_dlp/YoutubeDL.py:1605: in wrapper
    return func(self, *args, **kwargs)
yt_dlp/YoutubeDL.py:1740: in __extract_info
    ie_result = ie.extract(url)
yt_dlp/extractor/common.py:740: in extract
    ie_result = self._real_extract(url)
yt_dlp/extractor/neteasemusic.py:563: in _real_extract
    formats = self._extract_formats(info['mainSong'])
yt_dlp/extractor/neteasemusic.py:94: in _extract_formats
    self.raise_geo_restricted(
yt_dlp/extractor/common.py:1254: in raise_geo_restricted
    raise GeoRestrictedError(msg, countries=countries)
E   yt_dlp.utils.GeoRestrictedError: [netease:program] 10109055: No media links found; possibly due to geo restriction

During handling of the above exception, another exception occurred:
test/test_download.py:294: in test_template
    getattr(self, test_name)()
test/test_download.py:169: in test_template
    res_dict = ydl.extract_info(
yt_dlp/YoutubeDL.py:1594: in extract_info
    return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
yt_dlp/YoutubeDL.py:1621: in wrapper
    self.report_error(msg)
yt_dlp/YoutubeDL.py:1072: in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
yt_dlp/YoutubeDL.py:1011: in trouble
    raise DownloadError(message, exc_info)
E   yt_dlp.utils.DownloadError: ERROR: [netease:program] 10109055: No media links found; possibly due to geo restriction
E   This video is available in China.
E   You might want to use a VPN or a proxy server (with --proxy) to workaround.
----------------------------- Captured stdout call -----------------------------
[debug] Loaded 1823 extractors
[netease:program] Extracting URL: http://music.163.com/#/program?id=10109055
[netease:program] 10109055: Downloading program info
[netease:program] 32593346: Downloading song URL info: level standard
ERROR: [netease:program] 10109055: No media links found; possibly due to geo restriction
This video is available in China.
You might want to use a VPN or a proxy server (with --proxy) to workaround.
Traceback (most recent call last):
  File "/tmp/yt-dlp/yt_dlp/YoutubeDL.py", line 1605, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/yt-dlp/yt_dlp/YoutubeDL.py", line 1740, in __extract_info
    ie_result = ie.extract(url)
                ^^^^^^^^^^^^^^^
  File "/tmp/yt-dlp/yt_dlp/extractor/common.py", line 740, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/yt-dlp/yt_dlp/extractor/neteasemusic.py", line 563, in _real_extract
    formats = self._extract_formats(info['mainSong'])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/yt-dlp/yt_dlp/extractor/neteasemusic.py", line 94, in _extract_formats
    self.raise_geo_restricted(
  File "/tmp/yt-dlp/yt_dlp/extractor/common.py", line 1254, in raise_geo_restricted
    raise GeoRestrictedError(msg, countries=countries)
yt_dlp.utils.GeoRestrictedError: [netease:program] 10109055: No media links found; possibly due to geo restriction

=========================== short test summary info ============================
FAILED test/test_download.py::TestDownload::test_NetEaseMusicProgram_all - yt...
============================== 1 failed in 3.60s ===============================
```

```
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicDjRadio_all']
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0 -- /home/hafeoz/.local/share/hatch/env/virtual/yt-dlp/hPV3QzYK/hatch-test.py3.12/bin/python3
cachedir: .pytest_cache
rootdir: /tmp/yt-dlp
configfile: pyproject.toml
plugins: rerunfailures-14.0, xdist-3.6.1
collecting ... collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicDjRadio_all PASSED [100%]

============================== 1 passed in 3.65s ===============================
```

</details>

Following is a quick patch if we want to address it in this PR:

```diff
diff --git a/yt_dlp/extractor/neteasemusic.py b/yt_dlp/extractor/neteasemusic.py
index 0399efb1c..4c1cda555 100644
--- a/yt_dlp/extractor/neteasemusic.py
+++ b/yt_dlp/extractor/neteasemusic.py
@@ -291,7 +291,7 @@ class NetEaseMusicAlbumIE(NetEaseMusicBaseIE):
             'description': '桃几2021年翻唱合集',
             'thumbnail': r're:^http.*\.jpg',
         },
-        'playlist_mincount': 13,
+        'playlist_mincount': 12,
     }, {
         'url': 'http://music.163.com/#/album?id=220780',
         'info_dict': {
@@ -442,7 +442,7 @@ class NetEaseMusicMvIE(NetEaseMusicBaseIE):
             'ext': 'mp4',
             'title': '交换余生',
             'description': 'md5:e845872cff28820642a2b02eda428fea',
-            'creator': '林俊杰',
+            'creators': ['林俊杰'],
             'upload_date': '20200916',
             'thumbnail': r're:http.*\.jpg',
             'duration': 364,
@@ -457,7 +457,7 @@ class NetEaseMusicMvIE(NetEaseMusicBaseIE):
             'ext': 'mp4',
             'title': '이럴거면 그러지말지',
             'description': '白雅言自作曲唱甜蜜爱情',
-            'creator': '白娥娟',
+            'creators': ['白娥娟'],
             'upload_date': '20150520',
             'thumbnail': r're:http.*\.jpg',
             'duration': 216,
@@ -465,6 +465,7 @@ class NetEaseMusicMvIE(NetEaseMusicBaseIE):
             'like_count': int,
             'comment_count': int,
         },
+        'skip': 'Blocked outside Mainland China',
     }]
 
     def _real_extract(self, url):
@@ -506,12 +507,13 @@ class NetEaseMusicProgramIE(NetEaseMusicBaseIE):
             'ext': 'mp3',
             'title': '不丹足球背后的故事',
             'description': '喜马拉雅人的足球梦 ...',
-            'creator': '大话西藏',
+            'creators': ['大话西藏'],
             'timestamp': 1434179287,
             'upload_date': '20150613',
             'thumbnail': r're:http.*\.jpg',
             'duration': 900,
         },
+        'skip': 'Blocked outside Mainland China',
     }, {
         'note': 'This program has accompanying songs.',
         'url': 'http://music.163.com/#/program?id=10141022',
@@ -519,12 +521,13 @@ class NetEaseMusicProgramIE(NetEaseMusicBaseIE):
             'id': '10141022',
             'title': '滚滚电台的有声节目',
             'description': 'md5:8d594db46cc3e6509107ede70a4aaa3b',
-            'creator': '滚滚电台ORZ',
+            'creators': ['滚滚电台ORZ'],
             'timestamp': 1434450733,
             'upload_date': '20150616',
             'thumbnail': r're:http.*\.jpg',
         },
         'playlist_count': 4,
+        'skip': 'Blocked outside Mainland China',
     }, {
         'note': 'This program has accompanying songs.',
         'url': 'http://music.163.com/#/program?id=10141022',
@@ -533,7 +536,7 @@ class NetEaseMusicProgramIE(NetEaseMusicBaseIE):
             'ext': 'mp3',
             'title': '滚滚电台的有声节目',
             'description': 'md5:8d594db46cc3e6509107ede70a4aaa3b',
-            'creator': '滚滚电台ORZ',
+            'creators': ['滚滚电台ORZ'],
             'timestamp': 1434450733,
             'upload_date': '20150616',
             'thumbnail': r're:http.*\.jpg',
@@ -542,6 +545,7 @@ class NetEaseMusicProgramIE(NetEaseMusicBaseIE):
         'params': {
             'noplaylist': True,
         },
+        'skip': 'Blocked outside Mainland China',
     }]
 
     def _real_extract(self, url):
```

For some reason all "music programs" are unavailable for me both in browser and `yt-dlp` command line, so I'm unable to find suitable tests replacements.

timi137137 commented:
> @hafeoz @timi137137 please test my changes

<details>
<summary>Test results</summary>

```powershell
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusic_all']
========================== test session starts ===========================
platform win32 -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0 -- D:\App\Programme\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: D:\Data\Temp\yt-dlp, configfile: pyproject.toml
plugins: anyio-3.5.0
collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusic_all PASSED   [100%]

====================== 1 passed in 62.54s (0:01:02) ======================
```

---

**For this test I manually changed the array index**
```powershell
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicAlbum_all']
========================== test session starts ===========================
platform win32 -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0 -- D:\App\Programme\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: D:\Data\Temp\yt-dlp, configfile: pyproject.toml
plugins: anyio-3.5.0
collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicAlbum_all PASSED [100%]

=========================== 1 passed in 12.82s ===========================
```

---

```powershell
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicSinger_all']
========================== test session starts ===========================
platform win32 -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0 -- D:\App\Programme\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: D:\Data\Temp\yt-dlp, configfile: pyproject.toml
plugins: anyio-3.5.0
collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicSinger_all PASSED [100%]

=========================== 1 passed in 17.75s ===========================
```

---

```powershell
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicList_all']
========================== test session starts ===========================
platform win32 -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0 -- D:\App\Programme\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: D:\Data\Temp\yt-dlp, configfile: pyproject.toml
plugins: anyio-3.5.0
collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicList_all PASSED [100%]

=========================== 1 passed in 12.94s ===========================
```

---

**For this test I manually changed the creators**
```powershell
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicMv_all']
========================== test session starts ===========================
platform win32 -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0 -- D:\App\Programme\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: D:\Data\Temp\yt-dlp, configfile: pyproject.toml
plugins: anyio-3.5.0
collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicMv_all PASSED [100%]

=========================== 1 passed in 22.39s ===========================
```

---

**For this test I manually changed the creators**
```powershell
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicProgram_all']
========================== test session starts ===========================
platform win32 -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0 -- D:\App\Programme\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: D:\Data\Temp\yt-dlp, configfile: pyproject.toml
plugins: anyio-3.5.0
collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicProgram_all FAILED [100%]

================================ FAILURES ================================
_______________ TestDownload.test_NetEaseMusicProgram_all ________________
yt_dlp\YoutubeDL.py:1605: in wrapper
    return func(self, *args, **kwargs)
yt_dlp\YoutubeDL.py:1740: in __extract_info
    ie_result = ie.extract(url)
yt_dlp\extractor\common.py:740: in extract
    ie_result = self._real_extract(url)
yt_dlp\extractor\neteasemusic.py:563: in _real_extract
    formats = self._extract_formats(info['mainSong'])
yt_dlp\extractor\neteasemusic.py:94: in _extract_formats
    self.raise_geo_restricted(
yt_dlp\extractor\common.py:1254: in raise_geo_restricted
    raise GeoRestrictedError(msg, countries=countries)
E   yt_dlp.utils.GeoRestrictedError: [netease:program] 10109055: No media links found; possibly due to geo restriction

During handling of the above exception, another exception occurred:
test\test_download.py:294: in test_template
    getattr(self, test_name)()
test\test_download.py:169: in test_template
    res_dict = ydl.extract_info(
yt_dlp\YoutubeDL.py:1594: in extract_info
    return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
yt_dlp\YoutubeDL.py:1621: in wrapper
    self.report_error(msg)
yt_dlp\YoutubeDL.py:1072: in report_error
    self.trouble(f'{self._format_err("ERROR:", self.Styles.ERROR)} {message}', *args, **kwargs)
yt_dlp\YoutubeDL.py:1011: in trouble
    raise DownloadError(message, exc_info)
E   yt_dlp.utils.DownloadError: ERROR: [netease:program] 10109055: No media links found; possibly due to geo restriction
E   This video is available in China.
E   You might want to use a VPN or a proxy server (with --proxy) to workaround.
-------------------------- Captured stdout call --------------------------
[debug] Loaded 1823 extractors
[netease:program] Extracting URL: http://music.163.com/#/program?id=10109055
[netease:program] 10109055: Downloading program info
[netease:program] 32593346: Downloading song URL info: level standard
ERROR: [netease:program] 10109055: No media links found; possibly due to geo restriction
This video is available in China.
You might want to use a VPN or a proxy server (with --proxy) to workaround.
Traceback (most recent call last):
  File "D:\Data\Temp\yt-dlp\yt_dlp\YoutubeDL.py", line 1605, in wrapper
    return func(self, *args, **kwargs)
  File "D:\Data\Temp\yt-dlp\yt_dlp\YoutubeDL.py", line 1740, in __extract_info
    ie_result = ie.extract(url)
  File "D:\Data\Temp\yt-dlp\yt_dlp\extractor\common.py", line 740, in extract
    ie_result = self._real_extract(url)
  File "D:\Data\Temp\yt-dlp\yt_dlp\extractor\neteasemusic.py", line 563, in _real_extract
    formats = self._extract_formats(info['mainSong'])
  File "D:\Data\Temp\yt-dlp\yt_dlp\extractor\neteasemusic.py", line 94, in _extract_formats
    self.raise_geo_restricted(
  File "D:\Data\Temp\yt-dlp\yt_dlp\extractor\common.py", line 1254, in raise_geo_restricted
    raise GeoRestrictedError(msg, countries=countries)
yt_dlp.utils.GeoRestrictedError: [netease:program] 10109055: No media links found; possibly due to geo restriction

======================== short test summary info =========================
FAILED test/test_download.py::TestDownload::test_NetEaseMusicProgram_all
=========================== 1 failed in 13.74s ===========================
```

---

```powershell
Running ['pytest', '-Werror', '--tb=short', '-p', 'no:randomly', 'test/test_download.py::TestDownload::test_NetEaseMusicDjRadio_all']
========================== test session starts ===========================
platform win32 -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0 -- D:\App\Programme\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: D:\Data\Temp\yt-dlp, configfile: pyproject.toml
plugins: anyio-3.5.0
collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicDjRadio_all PASSED [100%]

=========================== 1 passed in 8.36s ============================
```
</details>

I modified some of the details and tested them according to hafeoz's guidance, and almost all of them passed
but, `test_NetEaseMusicProgram_all` is failed.I'm trying to figure out why

timi137137 commented:
The check is done, the problem is to determine whether it is geo.
The interface is fine and has normal returns.
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/a5dba8b0-fd3f-42f7-b2bb-0c133b199ad2)
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/14ce61b4-d452-434d-91a4-617de2faa027)

Summary: NetEaseMusicProgram cannot be analyzed in the same way

hafeoz commented:
All tests is running successfully on my machine now. Can anyone else confirm this?

I've also modified the `extract_formats` slightly so that duplicate audio URLs are ignored and only "real" sound qualities are shown. Can anyone else check if it's running as expected? For example instead of
```
ID       EXT RESOLUTION │ FILESIZE  TBR PROTO │ VCODEC     ACODEC  ABR
──────────────────────────────────────────────────────────────────────
standard mp3 audio only │  2.93MiB 128k http  │ audio only mp3    128k
exhigh   mp3 audio only │  7.33MiB 320k http  │ audio only mp3    320k
hires    mp3 audio only │  7.33MiB 320k http  │ audio only mp3    320k
jyeffect mp3 audio only │  7.33MiB 320k http  │ audio only mp3    320k
jymaster mp3 audio only │  7.33MiB 320k http  │ audio only mp3    320k
lossless mp3 audio only │  7.33MiB 320k http  │ audio only mp3    320k
sky      mp3 audio only │  7.33MiB 320k http  │ audio only mp3    320k
```
You should have
```
ID       EXT RESOLUTION │ FILESIZE  TBR PROTO │ VCODEC     ACODEC  ABR
──────────────────────────────────────────────────────────────────────
standard mp3 audio only │  2.93MiB 128k http  │ audio only mp3    128k
exhigh   mp3 audio only │  7.33MiB 320k http  │ audio only mp3    320k
```

timi137137 commented:
我一直都在，等我几分钟测试一下
Let me double confirm, pls wait

timi137137 commented:
All test is done,The result is all passed.
congratulate!

<details>
<summary>Test Random Song Result</summary>
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/6bb80576-4f90-4df2-8a64-19c71b06d675)
```ps
[info] Available formats for 29583265:
ID       EXT  RESOLUTION │   FILESIZE   TBR PROTO │ VCODEC     ACODEC   ABR
───────────────────────────────────────────────────────────────────────────
standard mp3  audio only │    3.93MiB  128k http  │ audio only mp3     128k
exhigh   mp3  audio only │    9.83MiB  320k http  │ audio only mp3     320k
lossless flac audio only │   23.33MiB  759k http  │ audio only flac    759k
jyeffect flac audio only │   85.17MiB 2773k http  │ audio only flac   2773k
jymaster flac audio only │  145.42MiB 4735k http  │ audio only flac   4735k
```

---

![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/2a557704-369b-44be-9715-f139b10d48a2)
```ps
[info] Available formats for 2134872913:
ID       EXT  RESOLUTION │   FILESIZE   TBR PROTO │ VCODEC     ACODEC   ABR
───────────────────────────────────────────────────────────────────────────
standard mp3  audio only │    3.27MiB  128k http  │ audio only mp3     128k
exhigh   mp3  audio only │    8.18MiB  320k http  │ audio only mp3     320k
lossless flac audio only │   18.94MiB  740k http  │ audio only flac    740k
hires    flac audio only │   38.31MiB 1498k http  │ audio only flac   1498k
jyeffect flac audio only │   68.09MiB 2663k http  │ audio only flac   2663k
jymaster flac audio only │  106.27MiB 4157k http  │ audio only flac   4157k
```

</details>

bashonly commented on yt_dlp/extractor/neteasemusic.py:
```diff
@@ -66,45 +66,40 @@ def _download_eapi_json(self, path, video_id, query_body, headers={}, **kwargs):
-            if not details:
+        for level in self._LEVELS:
+            song = traverse_obj(self._call_player_api(song_id, level), ('data', 0, {dict})) or {}
+            if song.get('level') in [song_format.get('level') for song_format in formats]:
+                break  # API returns existing level - the user have no access to further ones
```
this is wrong. there is no `level` field in the format dicts

hafeoz replied:
oof you're right. Fixed.

bashonly commented on yt_dlp/extractor/neteasemusic.py:
```diff
@@ -66,45 +66,40 @@ def _download_eapi_json(self, path, video_id, query_body, headers={}, **kwargs):
-                elif err == 0:
-                    err = traverse_obj(song, ('code', {int})) or 0
-
+            if 'md5' in song and song['md5'] in format_MD5s:
+                continue  # API returns existing audio file - this level is the same as existing ones
```
if we already `break` upon reaching a dupe, when will this code path be reached?

bashonly replied:
the idea with breaking the loop is so that we don't make all 7 api calls unnecessarily

hafeoz replied:
Yeah I'm probably not in the best mind when reasoning the logic while writing this... sorry. Reverted.

timi137137 replied:
the company was annoying, and he didn't have an interface that could directly display all the sound quality options.
The official client also uses traversal.I just went to decompile it

bashonly commented:
could you explain why you made the changes since my last push?

bashonly commented:
> could you explain why you made the changes since my last push?

specifically "fix early return if API returns no level information" -- is there an example where this actually happens?

hafeoz commented:
> could you explain why you made the changes since my last push?

So basically I've tried to make sure all tests passes.

The first commit ([fix early return if API returns no level information](https://github.com/yt-dlp/yt-dlp/pull/10258/commits/d80984c4ed5b907e5b1deb4b04a807d1bb655116)) is made to address [`NetEaseMusicProgram` test failing](https://github.com/yt-dlp/yt-dlp/pull/10258#issuecomment-2191241998) due to missing `level` key in API return.

The next commit ([detect duplicate song levels](https://github.com/yt-dlp/yt-dlp/pull/10258/commits/aecf4c04063b6021117bf0944e9437dca145decc)) is committed when I can't make sense of previous commit's extraction result (because there's a bug...). I've reverted this commit.

The next commit ([fix playlist test count](https://github.com/yt-dlp/yt-dlp/pull/10258/commits/3e25b02e5beb17f526dd94d35c3606c3b12a7007)) is due to [the test album](https://music.163.com/#/album?id=133153666) being updated and currently don't have the specified count anymore.

The next commit ([extract creators array instead of concating them](https://github.com/yt-dlp/yt-dlp/pull/10258/commits/bd9bb3bbf9db6ca529c01fd1cb3aa7e5177b8697)) changes `creator` to `creators` since the `info_dict` returns `creators` now.

The [use correct field in info_dict](https://github.com/yt-dlp/yt-dlp/pull/10258/commits/38ee4af64441d3f33b908d2e9b679e701c8ebc8c) commit fixes the first commit.

Finally the [Revert "detect duplicate song levels"](https://github.com/yt-dlp/yt-dlp/pull/10258/commits/29ad93b862c096ddd3bab9727b261dbafcd095f0) reverts the second commit.

hafeoz commented:
> specifically "fix early return if API returns no level information" -- is there an example where this actually happens?

One example would be `https://music.163.com/#/song?id=32593346`. The URL is "404 not found" via browser but `http://music.163.com/#/program?id=10109055` actually links to it. That's why the `NetEaseMusicProgram` was failing.

bashonly commented:
Thanks for the explanation. I've pushed a slight refactor to account for the netease:program behavior, let me know what you think

dirkf commented on yt_dlp/extractor/neteasemusic.py:
```diff
@@ -76,8 +76,8 @@ def _extract_formats(self, info):
-            if song.get('level') != level:
-                break  # We have already extracted the highest level the user has access to
+            if song.get('level') in [song_format.get('level') for song_format in formats]:
+                break  # API returns existing level - the user have no access to further ones
```
Maybe better to maintain a `set`, say `seen_levels`, and check for `song.get('level') in seen_levels`?

hafeoz replied:
Yes that's probably better, although I think [bashonly's commit](https://github.com/yt-dlp/yt-dlp/pull/10258/commits/009d2cc42f932e54554d02e54af2d4d325e383ad) should be enough

dirkf commented on yt_dlp/extractor/neteasemusic.py:
```diff
@@ -66,45 +66,39 @@ def _download_eapi_json(self, path, video_id, query_body, headers={}, **kwargs):
-                    err = traverse_obj(song, ('code', {int})) or 0
-
+        for level in self._LEVELS:
+            song = traverse_obj(self._call_player_api(song_id, level), ('data', 0, {dict})) or {}
```
Is there ever anything in `['data'][i]` for `i > 0`?

timi137137 replied:
data.length alawys is 0（At least for now）

hafeoz commented:
@timi137137 Can you check if high quality audio (using e.g. [the URL above](https://github.com/yt-dlp/yt-dlp/pull/10258#issuecomment-2191976414)) can be extracted? Especially the `sky` and `jymaster` one. I'm a bit confused about the fact that `jyeffect` and `sky` returns same bitrate audio while `jyeffect` returns a better one.

你能随便试一个音乐再确定一下`jymaster`还可以获取到吗？我理了一下头绪，感觉好像不能简单的用`actual_level != level`确认最高音乐质量...

如果现在获取不到`jymaster`音质了，可以发一下应用了下面patch之后的输出吗？我的账号没有会员好像什么高级音质都获取不了:
```patch
diff --git a/yt_dlp/extractor/neteasemusic.py b/yt_dlp/extractor/neteasemusic.py
index 929245756..3f6753631 100644
--- a/yt_dlp/extractor/neteasemusic.py
+++ b/yt_dlp/extractor/neteasemusic.py
@@ -80,7 +80,8 @@ def _extract_formats(self, info):
                 break  # Media is not available due to removal or geo-restriction
             actual_level = song.get('level')
             if actual_level and actual_level != level:
-                break  # We have already extracted the highest level the user has access to
+                print(actual_level)
+                #break  # We have already extracted the highest level the user has access to
             formats.append({
                 'url': song['url'],
                 'format_id': level,
```

timi137137 commented:
@hafeoz 好你稍等我一下，我刚刚看了一遍客户端抓包，新版本3.0的他改了请求看着就头疼。旧的还能用真的是谢天谢地
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/704771a8-6f41-483b-a628-46eba8ffbabb)

timi137137 commented:
@hafeoz
jymaster被错误过滤了，之前的版本能获取。 
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/33ef2854-7a43-4b7f-a198-f0fc31910602)

箭头指的是你让我patch以后得输出
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/6fdd0ba2-5ca7-4700-a503-7a28ed3bbf40)

hafeoz commented:
> @hafeoz jymaster被错误过滤了，之前的版本能获取。 ![image](https://private-user-images.githubusercontent.com/36812439/343314527-33ef2854-7a43-4b7f-a198-f0fc31910602.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk0MzI0NjksIm5iZiI6MTcxOTQzMjE2OSwicGF0aCI6Ii8zNjgxMjQzOS8zNDMzMTQ1MjctMzNlZjI4NTQtN2E0My00YjdmLWExOTgtZjBmYzMxOTEwNjAyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI2VDIwMDI0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmNjNiYzM4M2I4YWEyNzJmNTExYzg1NmNmYzE4MGIwYTVmZDAzM2Y2NjkzZjRhN2I1YjY4YWI2Njk3MjAwNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WElKXsoHE3HZ2uWI0H9TEmvxD_RadeYOsCO_WlSFxGY)

OK so we can't actually break the loop early if `level` is repeating. For some reason the API may return the same `level` even if there are more high-quality ones available. For example in this image the server returns the same `real_level` for the `jyeffect` and `sky` requests, while returning a completely different one for `jymaster`...

hafeoz commented:
@bashonly I've pushed a commit to `continue` instead of `break` when `actual_level != level`. This is because the API may return repeating levels even if higher level ones are actually available for the user (see image above). This will re-increase the number of requests but I can't think of a better solution for now.

timi137137 commented:
@hafeoz  目前市面上基本上都只有v3版本的接口，也就是我们现在在用的这个。
我刚刚尝试破解v6版本的接口但失败了。根据多次请求可以测试出v6版本的接口可以一个请求获取所有细节
只能说这个就静等大佬动手了，我是干不动

bashonly commented:
Pushed some more changes:
- added `higher` level (192kbps)
- improved extraction loop to (hopefully) break only when the user doesn't have access to the next tier (VIP or SVIP)

@timi137137 could you please try some links with `-vF` to see if the SVIP formats are properly extracted?

timi137137 commented:
@bashonly I tested two of them. They look good

<details>
<summary>Test Result</summary>

![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/1a51fbe2-8890-4310-a269-22fe1753abb2)

```ps
[netease:song] Extracting URL: https://music.163.com/song?id=2156449786&userid=304494550
[netease:song] 2156449786: Downloading song info
[netease:song] 2156449786: Downloading song URL info: level standard
[netease:song] 2156449786: Downloading song URL info: level higher
[netease:song] 2156449786: Downloading song URL info: level exhigh
[netease:song] 2156449786: Downloading song URL info: level lossless
[netease:song] 2156449786: Downloading song URL info: level hires
[netease:song] 2156449786: Downloading song URL info: level jyeffect
[netease:song] 2156449786: Downloading song URL info: level jymaster
[netease:song] 2156449786: Downloading song URL info: level sky
[netease:song] 2156449786: Downloading lyrics data
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 2156449786:
ID       EXT  RESOLUTION │   FILESIZE   TBR PROTO │ VCODEC     ACODEC   ABR
───────────────────────────────────────────────────────────────────────────
standard mp3  audio only │    2.82MiB  128k http  │ audio only mp3     128k
higher   mp3  audio only │    4.22MiB  192k http  │ audio only mp3     192k
exhigh   mp3  audio only │    7.04MiB  320k http  │ audio only mp3     320k
lossless flac audio only │   25.58MiB 1163k http  │ audio only flac   1163k
hires    flac audio only │   42.51MiB 1932k http  │ audio only flac   1932k
jyeffect flac audio only │   72.09MiB 3277k http  │ audio only flac   3277k
jymaster flac audio only │  126.82MiB 5766k http  │ audio only flac   5766k
```

---

![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/84d2fd13-b05d-49b8-b95f-d3e11cdc459c)

```ps
[netease:song] Extracting URL: https://music.163.com/song?id=30148963&userid=304494550
[netease:song] 30148963: Downloading song info
[netease:song] 30148963: Downloading song URL info: level standard
[netease:song] 30148963: Downloading song URL info: level higher
[netease:song] 30148963: Downloading song URL info: level exhigh
[netease:song] 30148963: Downloading song URL info: level lossless
[netease:song] 30148963: Downloading song URL info: level hires
[netease:song] 30148963: Downloading song URL info: level jyeffect
[netease:song] 30148963: Downloading song URL info: level jymaster
[netease:song] 30148963: Downloading song URL info: level sky
[netease:song] 30148963: Downloading lyrics data
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 30148963:
ID       EXT  RESOLUTION │   FILESIZE   TBR PROTO │ VCODEC     ACODEC   ABR
───────────────────────────────────────────────────────────────────────────
standard mp3  audio only │    3.82MiB  128k http  │ audio only mp3     128k
higher   mp3  audio only │    5.73MiB  192k http  │ audio only mp3     192k
exhigh   mp3  audio only │    9.55MiB  320k http  │ audio only mp3     320k
lossless flac audio only │   24.89MiB  833k http  │ audio only flac    833k
jyeffect flac audio only │   84.94MiB 2846k http  │ audio only flac   2846k
jymaster flac audio only │  141.44MiB 4739k http  │ audio only flac   4739k
```

</details>

dirkf commented:
If un-skipping `test_NetEaseMusicList_2`, I get this
```
AssertionError: None is not true : field title (value: '韩国Melon排行榜周榜') should match '韩国Melon排行榜周榜 [0-9]{4}-[0-9]{2}-[0-9]{2}'
```
It seems like this was expecting a `_live_title()`. Does this work in the Middle Kingdom?

timi137137 commented:
@dirkf Is this a direct call test? Using unit tests directly seems fine
```ps
pytest -Werror --tb=short -p no:randomly test/test_download.py::TestDownload::test_NetEaseMusicList_all
======================= test session starts =======================
platform win32 -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0 -- D:\App\Programme\anaconda3\python.exe
cachedir: .pytest_cache
rootdir: D:\Data\Temp\yt-dlp, configfile: pyproject.toml
plugins: anyio-3.5.0
collected 1 item

test/test_download.py::TestDownload::test_NetEaseMusicList_all PASSED [100%]

======================= 1 passed in 13.06s ========================
```

dirkf commented:
The API data sent to UK has this value
```console
(Pdb) pp info['playlist']['name']
'韩国Melon排行榜周榜'
```
So, without the date stamp. There was a test (http://music.163.com/#/discover/toplist?id=3733003) with a similar `title` pattern in the upstream code, so I don't think the timestamp is an artefact of _yt-dlp_.

If I change the `title` pattern
```diff
-            'title': 're:韩国Melon排行榜周榜 [0-9]{4}-[0-9]{2}-[0-9]{2})',
+            'title': 're:韩国Melon排行榜周榜(?: [0-9]{4}-[0-9]{2}-[0-9]{2})?$',
```
(**NB** a `re:` test pattern matches on the longest matching initial substring of a text value unless `$` is appended to force a full match), I get these missing attributes
```py
            'upload_date': '20200109',
            'timestamp': 1578569373,
            'uploader_id': '2937386',
            'tags': ['韩语', '榜单'],
            'uploader': 'Melon榜单',
```
and after adding those the test succeeds.

Perhaps there is a different page sent abroad that has a focused track (and so gets those extra values)?

dirkf commented:
Additionally, URLs like http://music.163.com/#/discover/toplist?id=3733003 (ie, with a `#` in a path component that doesn't signal a fragment) appear to contravene RFCs and also `urllib.parse` (and its predecessor) unless `allow_fragments=False` is passed. The correctly encoded URLs with `%23` seem to work and should be matched in `VALID_URL`s.

timi137137 commented:
Doesn't seem to be a problem?
![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/af85d75a-f986-4043-964d-8e237cece728)

bashonly commented:
Nobody is expected to run/update tests that are marked `skip`

dirkf commented:
If the test is marked as not working outside PRC but doesn't work there either, should it be fixed or removed?

timi137137 commented:
> Additionally, URLs like http://music.163.com/#/discover/toplist?id=3733003 (ie, with a `#` in a path component that doesn't signal a fragment) appear to contravene RFCs and also `urllib.parse` (and its predecessor) unless `allow_fragments=False` is passed. The correctly encoded URLs with `%23` seem to work and should be matched in `VALID_URL`s.

As far as front-end engineers are concerned, the # is normal

![image](https://github.com/yt-dlp/yt-dlp/assets/36812439/165c0f15-e935-4d70-8ce3-cd7e2afb9c77)

timi137137 commented:
I still do not understand what you are trying to do, our focus is not on the new API can not work?
Hasn't this regular expression proven itself to be no problem after all this time?
I hope that we can work together to promote the merger of the current PRs instead of entangling some problems that are not problematic
thx bro
